### DISCUSSION TITLE
Store selected tenant in database

### DISCUSSION THREAD

**Discussion by wq9578** [filamentphp/filament#10589] [Ideas]
When using multi-tenancy, the selected tenant should be stored in the database and used again when logging in (instead of using the alphabetically first tenant after login).
This more convenient approach is used in Jetstream, too.

**Comment by tembra** [comment]
The documentation has this covered and it's quite easy to implement it. You can check it at https://filamentphp.com/docs/3.x/panels/tenancy#setting-the-default-tenant

To save the latest tenant in a database you could use a tenant middleware to check when the current tenant is different of the stored one and then trigger the update.

**Comment by wq9578** [comment]
> The documentation has this covered and it's quite easy to implement it. You can check it at https://filamentphp.com/docs/3.x/panels/tenancy#setting-the-default-tenant

Thanks!
Works.

> To save the latest tenant in a database you could use a tenant middleware to check when the current tenant is different of the stored one and then trigger the update.

How can I do that (code example online)? I haven't written a middleware before.
Wouldn't it be better to implement a hook (when switching the tenant or when logging out)?

Generally, I would suggest making saving/restoring the tenant the default behaviour, as in Jetstream.
I meets the user's needs and provides a smooth user experience.

**Comment by wq9578** [comment]
I solved this using an event listener.
However, I noticed that the tenant is set about 50 times when changing the tenant in the panel.
This should be implemented more efficiently.

And generally speaking, saving/restoring the tenant should be the default behaviour.

```
class EventServiceProvider extends ServiceProvider
{
    /**
     * The event to listener mappings for the application.
     *
     * @var array<class-string, array<int, class-string>>
     */
    protected $listen = [
        Registered::class => [
            SendEmailVerificationNotification::class,
        ],
        TenantSet::class => [
            SaveTenant::class,
        ],
    ];
```

```
<?php

namespace App\Listeners;

use Filament\Events\TenantSet;
use Illuminate\Support\Facades\Log;

class SaveTenant
{
    /**
     * Create the event listener.
     */
    public function __construct()
    {
        //
    }

    /**
     * Handle the event.
     */
    public function handle(TenantSet $event): void
    {
        $user = $event->getUser();

        $user->current_team_id = $event->getTenant()->id;

        $user->save();

        Log::info('Tenant saved: ' . $event->getTenant()->id . ' (user ' . $user->id . ')');
    }
}
```

**Comment by wq9578** [comment]
So this should be the default behaviour (and implemented efficiently with only one event dispatched).
